blockchain/blocklocator.go

```diff
@@ -6,139 +6,94 @@ package blockchain
 // 	                              \-> 16a -> 17a
 //
 // The block locator for block 17a would be the hashes of blocks:
-// [17a 16a 15 14 13 12 11 10 9 8 6 2 genesis]
+// [17a 16a 15 14 13 12 11 10 9 8 7 6 4 genesis]
 type BlockLocator []*chainhash.Hash
 
-// blockLocatorFromHash returns a block locator for the passed block hash.
-// See BlockLocator for details on the algorithm used to create a block locator.
+// blockLocator returns a block locator for the passed block node.
 //
-// In addition to the general algorithm referenced above, there are a couple of
-// special cases which are handled:
-//
-//  - If the genesis hash is passed, there are no previous hashes to add and
-//    therefore the block locator will only consist of the genesis hash
-//  - If the passed hash is not currently known, the block locator will only
-//    consist of the passed hash
+// See BlockLocator for details on the algorithm used to create a block locator.
 //
 // This function MUST be called with the block index lock held (for reads).
-func (bi *blockIndex) blockLocatorFromHash(hash *chainhash.Hash) BlockLocator {
-	// The locator contains the requested hash at the very least.
-	locator := make(BlockLocator, 0, wire.MaxBlockLocatorsPerMsg)
-	locator = append(locator, hash)
-
-	// Nothing more to do if a locator for the genesis hash was requested.
-	if hash.IsEqual(bi.chainParams.GenesisHash) {
-		return locator
+func (bi *blockIndex) blockLocator(node *blockNode) BlockLocator {
+	if node == nil {
+		return nil
 	}
 
-	// Attempt to find the height of the block that corresponds to the
-	// passed hash, and if it's on a side chain, also find the height at
-	// which it forks from the main chain.
-	blockHeight := int32(-1)
-	forkHeight := int32(-1)
-	node, exists := bi.index[*hash]
-	if !exists {
-		// Try to look up the height for passed block hash.  Assume an
-		// error means it doesn't exist and just return the locator for
-		// the block itself.
-		var height int32
-		err := bi.db.View(func(dbTx database.Tx) error {
-			var err error
-			height, err = dbFetchHeightByHash(dbTx, hash)
-			return err
-		})
-		if err != nil {
-			return locator
-		}
-
-		blockHeight = height
+	// Calculate the max number of entries that will ultimately be in the
+	// block locator.  See the description of the algorithm for how these
+	// numbers are derived.
+	var maxEntries uint8
+	if node.height <= 12 {
+		maxEntries = uint8(node.height) + 1
 	} else {
-		blockHeight = node.height
-
-		// Find the height at which this node forks from the main chain
-		// if the node is on a side chain.
-		if !node.inMainChain {
-			for n := node; n.parent != nil; n = n.parent {
-				if n.inMainChain {
-					forkHeight = n.height
-					break
-				}
-			}
-		}
+		// Requested hash itself + previous 10 entries + genesis block.
+		// Then floor(log2(height-10)) entries for the skip portion.
+		adjustedHeight := uint32(node.height) - 10
+		maxEntries = 12 + fastLog2Floor(adjustedHeight)
```

Comment:
I've calculated `maxEntries` as 11 + floor(log_2(node.height - 9)), for height > 9. For example, if calculating the locator for node with height 12, the locator is 12, 11, ..., 2, 0, which has 12 elements, whereas `12 + floor(log(12 - 10))` would calculate it as 13.

Say node.height is `n` and you fill the first 10 elements of the block locator. The 10th element has height `n' = n - 9`. The remaining nodes in the locator are at heights `n' - 1`, `n' - 1 - 2`, `n' - 1 - 2 - 4`, `n' - 1 - 2 - 4 - 8`, ... or `n' + 1 - 2^(i+1)`. The number of such elements in this sequence of remaining nodes (after the first 10) is `ceil(log_2(n' + 1))` or `1 + floor(log_2(n'))` for `n' > 1`. Thus the total locator length `11 + floor(log_2(n - 9))`.